File: growth/oj/leet_code/algorithms/15-3sum.py
# -*- coding: utf-8 -*-


class Solution(object):
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        nums.sort()

        res = set()
        for i, a in enumerate(nums[:-2]): # 因为要3个数才构成一组解
        	l = i + 1
        	r = len(nums) - 1
        	while l < r:
        		tmp = nums[l] + nums[r]
        		if tmp == -a:
        			res.add((a, nums[l], nums[r]))
        			l += 1
        			r -= 1
        			while l < len(nums)-1 and nums[l] == nums[l-1] and r > 0 and nums[r] == nums[r+1]:
        				l += 1
        				r -= 1
        		elif tmp > -a:
        			r -= 1
        		else:
        			l += 1
        return map(list, res)


if __name__ == '__main__':
	obj = Solution()
	print obj.threeSum([-1, 0, 1, 2, -1, -4])
	print obj.threeSum([-1, 0, 1, -1, 0, 1])
	print obj.threeSum([-2,0,1,1,2])
	print obj.threeSum([-4,-2,-2,-2,0,1,2,2,2,3,3,4,4,6,6])

File: growth/oj/leet_code/algorithms/15-3sum.cpp
#include<iostream>

using namespace std;

class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> res;
        if (nums.size() < 3) return res;

        sort(nums.begin(), nums.end());

        for (int i = 0; i < nums.size() - 2; i++) {
            if (i == 0 || (i > 0 && nums[i-1] != nums[i])) {
                int target = -nums[i];
                int lo = i + 1, hi = nums.size() - 1;
                while(lo < hi) {
                    if (nums[lo] + nums[hi] == target) {
                        res.push_back(vector<int>{nums[i], nums[lo], nums[hi]});

                        while(lo < hi && nums[lo] == nums[lo+1]) lo++;
                        while(lo < hi && nums[hi] == nums[hi-1]) hi--;
                        lo++;
                        hi--;
                    } else if (nums[lo] + nums[hi] < target) {
                        lo++;
                    } else {
                        hi--;
                    }
                }
            }
        }
        return res;
    }
};


int main() {
    Solution s;
    return 0;
}
